### PR TITLE
asv bench cleanup - groupby

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Clean-up of the groupby benchmarks. 

This is an example of how they can be cleaned up (grouping benchmarks with the same setup in common classes, removing a lot of the setup functions in this way)

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@jreback @qwhelan suppose this is OK to merge? 
This was one of the larger files, but as you can see, quite some line reduction is possible with some manual clean-up (also makes it clearer to add benchmarks)

**Comment by jreback** [comment]
lgtm. I did sligthly different when I added some benchmarks [here](https://github.com/pydata/pandas/blob/master/asv_bench/benchmarks/binary_ops.py#L204)
as I subclassed. But both approaches make sense (and shorter boilerplate code)

**Comment by jreback** [comment]
a couple fail: ``asv dev --bench groupby``

```
[ 15.38%] Â·Â·Â· Running groupby.groupby_first_last.time_groupby_nth_float32_any                                                                                                                                                     fed
[ 15.38%] Â·Â·Â·Â·Â· Traceback (most recent call last):
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/asv-0.2.dev795+c27e60f7-py2.7.egg/asv/benchmark.py", line 774, in <module>
                    commands[mode](args)
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/asv-0.2.dev795+c27e60f7-py2.7.egg/asv/benchmark.py", line 751, in main_run
                    result = benchmark.do_run()
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/asv-0.2.dev795+c27e60f7-py2.7.egg/asv/benchmark.py", line 443, in do_run
                    return self.run(*self._current_params)
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/asv-0.2.dev795+c27e60f7-py2.7.egg/asv/benchmark.py", line 510, in run
                    timing = timer.timeit(number)
                  File "/Users/jreback/miniconda/lib/python2.7/timeit.py", line 201, in timeit
                    timing = self.inner(it, self.timer)
                  File "/Users/jreback/miniconda/lib/python2.7/timeit.py", line 100, in inner
                    _func()
                  File "/Users/jreback/pandas/asv_bench/benchmarks/groupby.py", line 64, in time_groupby_nth_float32_any
                    self.data2.groupby(self.labels).nth(0, dropna='all')
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/pandas/core/groupby.py", line 903, in nth
                    level=self.level, sort=self.sort)
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/pandas/core/groupby.py", line 2158, in _get_grouper
                    level=level, sort=sort, in_axis=in_axis)
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/pandas/core/groupby.py", line 1879, in __init__
                    self.grouper = _convert_grouper(index, grouper)
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/pandas/core/groupby.py", line 2185, in _convert_grouper
                    raise AssertionError('Grouper and axis must be same length')
                AssertionError: Grouper and axis must be same length


[ 16.92%] Â·Â·Â·Â·Â· Traceback (most recent call last):
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/asv-0.2.dev795+c27e60f7-py2.7.egg/asv/benchmark.py", line 774, in <module>
                    commands[mode](args)
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/asv-0.2.dev795+c27e60f7-py2.7.egg/asv/benchmark.py", line 751, in main_run
                    result = benchmark.do_run()
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/asv-0.2.dev795+c27e60f7-py2.7.egg/asv/benchmark.py", line 443, in do_run
                    return self.run(*self._current_params)
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/asv-0.2.dev795+c27e60f7-py2.7.egg/asv/benchmark.py", line 510, in run
                    timing = timer.timeit(number)
                  File "/Users/jreback/miniconda/lib/python2.7/timeit.py", line 201, in timeit
                    timing = self.inner(it, self.timer)
                  File "/Users/jreback/miniconda/lib/python2.7/timeit.py", line 100, in inner
                    _func()
                  File "/Users/jreback/pandas/asv_bench/benchmarks/groupby.py", line 70, in time_groupby_nth_float64_any
                    self.data.groupby(self.labels).nth(0, dropna='all')
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/pandas/core/groupby.py", line 903, in nth
                    level=self.level, sort=self.sort)
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/pandas/core/groupby.py", line 2158, in _get_grouper
                    level=level, sort=sort, in_axis=in_axis)
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/pandas/core/groupby.py", line 1879, in __init__
                    self.grouper = _convert_grouper(index, grouper)
                  File "/Users/jreback/miniconda/lib/python2.7/site-packages/pandas/core/groupby.py", line 2185, in _convert_grouper
                    raise AssertionError('Grouper and axis must be same length')
                AssertionError: Grouper and axis must be same length
```

**Comment by jorisvandenbossche** [comment]
yes, I know, but I didn't yet address that here (only restructuring the file, not fixing the tests themselves)

The subclassing is also fine, but there you had the same test function but different creation setup, so in that case it's more awkward to put it in one test class

**Comment by jorisvandenbossche** [comment]
although, probably even the easier option is to use ``params`` in this case, as only the value for `N` changes, http://asv.readthedocs.org/en/latest/writing_benchmarks.html#parameterized-benchmarks

**Comment by jreback** [comment]
oh, all these cool features. Ok, once the suite is cleaned up prob need to put a some docs in place when/how to do various things.

**Comment by qwhelan** [comment]
It'll be a few hours until I can take a look (onsite + bunch of meetings
this evening), but I'm all for refactoring provided we prevent future
changes to the vbench groupby benchmarks (delete/comment out/scary warning).

**Comment by jorisvandenbossche** [comment]
Yes, I think we should certainly ensure we don't change the vbenches anymore. Otherwise it has no point in starting to adapt the asv benches.

Will do a PR with a notice. I think we can leave them in for some time for those we didn't switch yet

File: asv_bench/benchmarks/groupby.py [hunk 1]
```diff
@@ -1535,7 +674,9 @@ def setup(self):
         self.secid_max = int('F0000000', 16)
         self.step = ((self.secid_max - self.secid_min) // (self.n_securities - 1))
         self.security_ids = map((lambda x: hex(x)[2:10].upper()), range(self.secid_min, (self.secid_max + 1), self.step))
-        self.data_index = MultiIndex(levels=[self.dates.values, self.security_ids], labels=[[i for i in range(self.n_dates) for _ in xrange(self.n_securities)], (range(self.n_securities) * self.n_dates)], names=['date', 'security_id'])
+        self.data_index = MultiIndex(levels=[self.dates.values, self.security_ids],
+                                     labels=[[i for i in range(self.n_dates) for _ in xrange(self.n_securities)], (range(self.n_securities) * self.n_dates)],
```
Comment (qwhelan) [inline asv_bench/benchmarks/groupby.py]: @jorisvandenbossche  Looks like I missed this in the first pass, but this ```xrange``` call will break things on Py3. Compatability is being handled by overwriting ```range``` in ```pandas_vb_common```.

**Comment by qwhelan** [comment]
@jorisvandenbossche Other than the ```xrange``` call that's my fault, looks good to me.

**Comment by jreback** [comment]
FYI, I have run benchmarks since 0.14 (not exactly on the tag as not a real easy way to do this). and pushed to gh-pages

http://pydata.github.io/pandas/

**Comment by qwhelan** [comment]
Yeah, I've got some local patches to fix the run-against-tag issues that
result from trying to linearize history by using the first parent. I'll
chat with the asv devs to see if they're dead-set on that or whether that
can be configurable.

File: asv_bench/benchmarks/groupby.py (hunk 1 above)
Comment (jorisvandenbossche) [inline asv_bench/benchmarks/groupby.py]: thanks, fixed!

**Comment by jorisvandenbossche** [comment]
@jreback there seems to be some regressions since the last release (eg http://pydata.github.io/pandas/#plotting.plot_timeseries_period.time_plot_timeseries_period) maybe start listing them in the 0.17 release issue?

**Comment by jorisvandenbossche** [comment]
@jreback @qwhelan another question: how do you handle the fact that some tests will fail on older pandas versions? (with vbench there was the start date you could give with the benchmark)

**Comment by jreback** [comment]
@jorisvandenbossche I put that one on the 0.17.0 RLS issue. How do you 'find' regressions though (e.g. except for looking at charts directly). Did I not do something?

**Comment by qwhelan** [comment]
Raising NotImplementedError is equivalent to SkipTest. Probably best done
in setup().

**Comment by qwhelan** [comment]
In the "all benchmarks" view, there's a "Regressions" button in the upper
left. It's not spitting out anything due the tag issue I mentioned earlier.

**Comment by jreback** [comment]
ahh I c

ok I did it with the mark I eyeball :)

**Comment by jorisvandenbossche** [comment]
@jreback In Christophers version it works, so you can see there how it looks: http://qwhelan.github.io/pandas_asv/#regressions?sort=3&dir=desc

But for checking a release, I think the `asv continuous` between the previous release and the current release is also an good option to have an overview of possible regressions

**Comment by jreback** [comment]
@qwhelan also maybe I am using it wrongly

but asv gh-pages crashed for me

I just did: asv publish

the copied the html dir

**Comment by qwhelan** [comment]
No, the regression page not working is definitely a bug in master. Also, I
did not use asv gh-pages (didn't realize it was there). I did the pages
repo setup manually but I can investigate the crash tomorrow.

**Pull request merged**